src/plugins/liveobjects/livemap.ts

```diff
@@ -53,18 +92,197 @@ export class LiveMap extends LiveObject<LiveMapData> {
+        50000,
+        500,
+      );
+    }
+
+    switch (op.action) {
+      case StateOperationAction.MAP_CREATE:
+        if (this._client.Utils.isNil(op.map)) {
+          this._throwNoPayloadError(op);
+        } else {
+          this._applyMapCreate(op.map);
+        }
+        break;
+
+      case StateOperationAction.MAP_SET:
+        if (this._client.Utils.isNil(op.mapOp)) {
+          this._throwNoPayloadError(op);
+        } else {
+          this._applyMapSet(op.mapOp, msg.serial);
+        }
+        break;
+
+      case StateOperationAction.MAP_REMOVE:
+        if (this._client.Utils.isNil(op.mapOp)) {
+          this._throwNoPayloadError(op);
+        } else {
+          this._applyMapRemove(op.mapOp, msg.serial);
+        }
+        break;
+
+      default:
+        throw new this._client.ErrorInfo(
+          `Invalid ${op.action} op for LiveMap objectId=${this.getObjectId()}`,
+          50000,
+          500,
+        );
+    }
   }
 
   protected _getZeroValueData(): LiveMapData {
     return { data: new Map<string, MapEntry>() };
   }
+
+  private _throwNoPayloadError(op: StateOperation): void {
+    throw new this._client.ErrorInfo(
+      `No payload found for ${op.action} op for LiveMap objectId=${this.getObjectId()}`,
+      50000,
+      500,
+    );
+  }
+
+  private _applyMapCreate(op: StateMap): void {
+    if (this._semantics !== op.semantics) {
+      throw new this._client.ErrorInfo(
+        `Cannot apply MAP_CREATE op on LiveMap objectId=${this.getObjectId()}; map's semantics=${this._semantics}, but op expected ${op.semantics}`,
+        50000,
+        500,
+      );
+    }
+
+    // in order to apply MAP_CREATE op for an existing map, we should merge their underlying entries keys.
+    // we can do this by iterating over entries from MAP_CREATE op and apply changes on per-key basis as if we had MAP_SET, MAP_REMOVE operations.
+    Object.entries(op.entries ?? {}).forEach(([key, entry]) => {
+      // for MAP_CREATE op we must use dedicated timeserial field available on an entry, instead of a timeserial on a message
+      const opOriginTimeserial = entry.timeserial;
+      if (entry.tombstone === true) {
+        // entry in MAP_CREATE op is deleted, try to apply MAP_REMOVE op
+        this._applyMapRemove({ key }, opOriginTimeserial);
+      } else {
+        // entry in MAP_CREATE op is not deleted, try to set it via MAP_SET op
+        this._applyMapSet({ key, data: entry.data }, opOriginTimeserial);
+      }
+    });
+  }
+
+  private _applyMapSet(op: StateMapOp, opOriginTimeserialStr: string | undefined): void {
+    const { ErrorInfo, Utils } = this._client;
+
+    const opTimeserial = DefaultTimeserial.calculateTimeserial(this._client, opOriginTimeserialStr);
+    const existingEntry = this._dataRef.data.get(op.key);
+    if (
+      existingEntry &&
+      (opTimeserial.before(existingEntry.timeserial) || opTimeserial.equal(existingEntry.timeserial))
```

Comment:
On the realtime side, we apply the op if the timeserial >= the entry timeserial: https://github.com/ably/realtime/blob/main/go/realtime/lib/state/map.go#L135

It shouldn't be possible to generate timeserial that == the entry timeserial, but we should align the implementation on both sides.

Reply:
oh, no problem, will change that.

Reply:
made suggested changes, see compare https://github.com/ably/ably-js/compare/0a70bb1faa7145bcef36018758bd921069609c2d..a9f916939a12752f478c21972a08dd732bf729ac